In [1]:
from dotenv import load_dotenv
from langchain_openai import ChatOpenAI

load_dotenv()

llm = ChatOpenAI(model="gpt-4o", temperature=0)

## LLM 보고서 생성하기

In [2]:
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.output_parsers  import StrOutputParser

prompt = ChatPromptTemplate.from_messages([
    ("system", """
     Want assistance provided by qualified individuals enabled with experience on understanding charts 
     using technical analysis tools while interpreting macroeconomic environment prevailing across world consequently 
     assisting customers acquire long term advantages requires clear verdicts therefore seeking same through informed 
     predictions written down precisely! First statement contains following content- 
     “Can you tell us what future stock market looks like based upon current conditions ?"."""),
    ("user", """
        {company} 주식에 투자를 해도 될까요? 마크다운 형식의 투자 보고서를 한글로 작성해 주세요.
     """)
])

output_parser = StrOutputParser()

chain = prompt | llm | output_parser

company = "Microsoft"

response= chain.invoke({"company": company})
print(response)


# 마이크로소프트(Microsoft) 주식 투자 보고서

## 1. 회사 개요
마이크로소프트(Microsoft)는 세계 최대의 소프트웨어 회사 중 하나로, 운영 체제, 생산성 소프트웨어, 클라우드 서비스, 인공지능, 게임 등 다양한 분야에서 사업을 운영하고 있습니다. 특히, 클라우드 컴퓨팅 서비스인 Azure와 생산성 소프트웨어인 Office 365는 회사의 주요 수익원 중 하나입니다.

## 2. 최근 주가 동향
마이크로소프트의 주가는 최근 몇 년간 꾸준한 상승세를 보였습니다. 특히, 클라우드 서비스의 성장과 인공지능 기술의 발전이 주가 상승에 긍정적인 영향을 미쳤습니다. 그러나, 최근 글로벌 경제 불확실성과 기술주에 대한 조정으로 인해 변동성이 증가할 가능성도 있습니다.

## 3. 재무 분석
- **매출 성장**: 마이크로소프트는 지속적인 매출 성장을 기록하고 있으며, 특히 클라우드 부문에서의 성장이 두드러집니다.
- **수익성**: 높은 영업 이익률과 순이익률을 유지하고 있으며, 이는 효율적인 비용 관리와 고부가가치 제품의 판매 덕분입니다.
- **현금 흐름**: 강력한 현금 흐름을 보유하고 있어, 배당금 지급 및 주식 환매 프로그램을 통해 주주 가치를 제고하고 있습니다.

## 4. 시장 및 경쟁 환경
마이크로소프트는 Amazon, Google, Apple 등과 같은 대형 기술 기업들과 경쟁하고 있습니다. 특히, 클라우드 시장에서는 Amazon의 AWS와 치열한 경쟁을 벌이고 있습니다. 그러나, 마이크로소프트는 다양한 제품 포트폴리오와 강력한 브랜드 인지도를 바탕으로 경쟁 우위를 유지하고 있습니다.

## 5. 리스크 요인
- **글로벌 경제 불확실성**: 경제 침체나 금리 인상 등은 기술주 전반에 부정적인 영향을 미칠 수 있습니다.
- **규제 리스크**: 기술 기업에 대한 규제가 강화될 경우, 사업 운영에 영향을 미칠 수 있습니다.
- **기술 변화**: 빠르게 변화하는 기술 환경에서 경쟁력을 유지하기 위한 지속적인 혁신이 필요합니다.

## 6

# 기본 정보 출력

In [3]:
import yfinance as yf

# Microsoft 주식 심볼 (MSFT)
ticker = 'MSFT'

# 티커 정보 가져오기
msft = yf.Ticker(ticker)

# 기본 정보 조회
basic_info = {
    "회사명": msft.info.get('longName'),
    "산업": msft.info.get('industry'),
    "부문": msft.info.get('sector'),
    "시가총액": msft.info.get('marketCap'),
    "총 발행 주식 수": msft.info.get('sharesOutstanding'),
    "본사 위치": msft.info.get('country'),
    "CEO": msft.info.get('ceo'),  # 일부 주식에서 사용 가능
    "웹사이트": msft.info.get('website'),
}

# 기본 정보 출력
for key, value in basic_info.items():
    print(f"{key}: {value}")

회사명: Microsoft Corporation
산업: Software - Infrastructure
부문: Technology
시가총액: 2782161797120
총 발행 주식 수: 7425629076
본사 위치: United States
CEO: None
웹사이트: https://www.microsoft.com


## 과거 주가데이터 정보

In [4]:
msft.history(period="1mo")

,Open,High,Low,Close,Volume,Dividends,Stock Splits
Date,,,,,,,
2026-03-09 00:00:00-04:00,404.920013,410.209991,403.500000,409.410004,30131900,0.0,0.0
2026-03-10 00:00:00-04:00,410.029999,410.200012,402.929993,405.760010,31706400,0.0,0.0
2026-03-11 00:00:00-04:00,405.570007,409.010010,401.589996,404.880005,25512100,0.0,0.0
2026-03-12 00:00:00-04:00,404.630005,406.119995,401.709991,401.859985,27263900,0.0,0.0
2026-03-13 00:00:00-04:00,401.000000,404.799988,394.250000,395.549988,26848000,0.0,0.0
2026-03-16 00:00:00-04:00,398.070007,400.630005,394.790009,399.950012,27733700,0.0,0.0
2026-03-17 00:00:00-04:00,400.269989,404.399994,397.750000,399.410004,26228300,0.0,0.0
2026-03-18 00:00:00-04:00,397.130005,398.000000,391.000000,391.790009,25908500,0.0,0.0
2026-03-19 00:00:00-04:00,390.100006,392.489990,387.059998,389.019989,25138800,0.0,0.0


### 재무제표 정보 가져오기

In [5]:
# 재무제표 정보 가져오기
income_statement = msft.financials  # 손익계산서
balance_sheet = msft.balance_sheet  # 대차대조표
cashflow_statement = msft.cashflow  # 현금흐름표

### 주요 지표 조회

In [6]:
# 주요 지표 조회
pe_ratio = msft.info['forwardPE']           # P/E 비율 (예상)
pb_ratio = msft.info['priceToBook']         # P/B 비율
eps = msft.info['trailingEps']              # EPS (주당순이익, 최근 12개월)
roe = msft.info['returnOnEquity']           # ROE (자기자본이익률)

# 데이터 출력
print(f"P/E Ratio (Forward): {pe_ratio}")
print(f"P/B Ratio: {pb_ratio}")
print(f"EPS (Trailing 12 Months): {eps}")
print(f"ROE: {roe}")

P/E Ratio (Forward): 19.860641
P/B Ratio: 7.1145105
EPS (Trailing 12 Months): 15.98
ROE: 0.34390998


### 배당수익률 조회

In [7]:
# 배당수익률 조회
dividend_yield = msft.info.get('dividendYield')

# 배당수익률 출력 (퍼센트로 변환)
if dividend_yield:
    dividend_yield_percent = dividend_yield * 100
    print(f"배당수익률: {dividend_yield_percent:.2f}%")
else:
    print("Microsoft는 현재 배당을 지급하지 않습니다.")

배당수익률: 97.00%


### 현재 거래량 조회 (최근 거래일 기준)

In [8]:
# 현재 거래량 조회 (최근 거래일 기준)
current_volume = msft.info.get('volume')

# 최근 거래량 출력
print(f"현재 거래량: {current_volume}")

현재 거래량: 32957404


In [9]:
# 과거 1개월 간의 주가 및 거래량 데이터 가져오기
historical_data = msft.history(period="1mo")

# 거래량 데이터 출력
print(historical_data['Volume'])

Date
2026-03-09 00:00:00-04:00    30131900
2026-03-10 00:00:00-04:00    31706400
2026-03-11 00:00:00-04:00    25512100
2026-03-12 00:00:00-04:00    27263900
2026-03-13 00:00:00-04:00    26848000
2026-03-16 00:00:00-04:00    27733700
2026-03-17 00:00:00-04:00    26228300
2026-03-18 00:00:00-04:00    25908500
2026-03-19 00:00:00-04:00    25138800
2026-03-20 00:00:00-04:00    50853200
2026-03-23 00:00:00-04:00    29680100
2026-03-24 00:00:00-04:00    42733600
2026-03-25 00:00:00-04:00    31181200
2026-03-26 00:00:00-04:00    36836600
2026-03-27 00:00:00-04:00    37883400
2026-03-30 00:00:00-04:00    44797000
2026-03-31 00:00:00-04:00    45244400
2026-04-01 00:00:00-04:00    29417200
2026-04-02 00:00:00-04:00    24099100
2026-04-06 00:00:00-04:00    16146600
2026-04-07 00:00:00-04:00    21443300
2026-04-08 00:00:00-04:00    32957404
Name: Volume, dtype: int64


In [ ]:
#종목 투자보고서 프롬프트 실행

from langchain_openai import ChatOpenAI
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.output_parsers  import StrOutputParser

llm = ChatOpenAI(model="gpt-4o-mini", temperature=0.3)

chat_template = ChatPromptTemplate.from_messages([
    ("system", """Want assistance provided by qualified individuals enabled with experience on understanding charts using technical analysis tools while interpreting macroeconomic environment prevailing across world consequently assisting customers acquire long term advantages requires clear verdicts therefore seeking same through informed predictions written down precisely! First statement contains following content- 'Can you tell us what future stock market looks like based upon current conditions ?'"""),
    ("human", """회사: {company}

                [기본정보]
                {basic_info}

                [분기 재무]
                {financials}

                요구사항:
                1) 사업 개요와 최근 분기 핵심 포인트 3가지
                2) 실적 추세(매출/영업이익/순이익) 요약
                3) 리스크 2가지, 모멘텀 2가지
                4) 투자 아이디어와 관찰지표 체크리스트
                5) 10줄 요약
                출력은 마크다운으로 작성"""
    )
])

chain = chat_template | llm | StrOutputParser()

def investment_report(company: str, basic_info: str, financials: str) -> str:
    
    # 1. chat_template에 변수를 주입하여 메시지 객체 생성
    # format_messages나 invoke를 사용합니다.
    prompt_params = {
        "company": company,
        "basic_info": basic_info,
        "financials": financials
    }
    
    # 2. llm에 완성된 prompt_value를 전달
    return chain.invoke(prompt_params)


In [12]:

from langchain_core.prompts import ChatPromptTemplate
from langchain_core.output_parsers import StrOutputParser
from stock_info.stock_info import Stock

prompt = ChatPromptTemplate.from_messages([
    ("system", """
     Want assistance provided by qualified individuals enabled with experience on understanding charts 
     using technical analysis tools while interpreting macroeconomic environment prevailing across world consequently 
     assisting customers acquire long term advantages requires clear verdicts therefore seeking same through informed 
     predictions written down precisely! First statement contains following content- 
     “Can you tell us what future stock market looks like based upon current conditions ?"."""),
    ("user", """
        {company} 주식에 투자를 해도 될까요? 
        아래의 기본정보, 재무제표를 참고해 마크다운 형식의 투자 보고서를 한글로 작성해 주세요.
        
        기본정보:
        {basic_info}
        
        재무제표:
        {financial_statement}
     """)
])

output_parser = StrOutputParser()

chain = prompt | llm | output_parser

company = "Apple"
symbol = "AAPL"

stock = Stock(symbol)

response= chain.invoke({
    "company": company,
    "basic_info": stock.get_basic_info(),
    "financial_statement": stock.get_financial_statement()})
print(response)

# Apple Inc. 투자 보고서

## 회사 개요
- **회사명**: Apple Inc.
- **산업**: 소비자 전자제품
- **섹터**: 기술
- **시가총액**: $3,805,292,789,760
- **발행 주식 수**: 14,681,140,000

## 재무 분석

### 1. 손익계산서 분석
Apple Inc.의 최근 5분기 동안의 손익계산서를 분석한 결과, 다음과 같은 주요 사항을 확인할 수 있습니다:

- **총 매출**: 2025년 12월 31일 기준으로 $143.756 billion으로, 전년 동기 대비 증가세를 보였습니다. 이는 Apple의 제품 및 서비스에 대한 지속적인 수요를 반영합니다.
- **영업 이익**: 2025년 12월 31일 기준으로 $50.852 billion으로, 매출 증가에 따라 영업 이익도 증가했습니다.
- **순이익**: 2025년 12월 31일 기준으로 $42.097 billion으로, 이는 전년 동기 대비 증가한 수치입니다.

### 2. 대차대조표 분석
대차대조표를 통해 Apple Inc.의 재무 건전성을 평가할 수 있습니다:

- **총 자산**: 2025년 12월 31일 기준으로 $379.297 billion으로, 자산이 꾸준히 증가하고 있습니다.
- **총 부채**: 2025년 12월 31일 기준으로 $291.107 billion으로, 부채 비율이 안정적으로 유지되고 있습니다.
- **주주 자본**: 2025년 12월 31일 기준으로 $88.19 billion으로, 주주 자본이 증가하고 있어 재무 건전성이 강화되고 있습니다.

### 3. 현금 흐름 분석
현금 흐름표를 통해 Apple Inc.의 현금 관리 상태를 평가할 수 있습니다:

- **영업 현금 흐름**: 2025년 12월 31일 기준으로 $53.925 billion으로, 강력한 영업 현금 흐름을 유지하고 있습니다.
- **투자 현금 흐름**: 2025년 12월 31일 기준으로 -$4.886 billion으로, 이는 주로 연구 개발 및 자산 취득

In [14]:
from stock_info.stock_info import Stock
from report_service.investment_report import investment_report

company = "Apple"
symbol = "AAPL"

stock = Stock(symbol)

basic_info = stock.get_basic_info()
financial_statement = stock.get_financial_statement()

In [15]:
basic_info

'| 항목              | Value                |\n|:------------------|:---------------------|\n| longName          | Apple Inc.           |\n| industry          | Consumer Electronics |\n| sector            | Technology           |\n| marketCap         | 3805292789760        |\n| sharesOutstanding | 14681140000          |'

| 항목              | Value                |
|:------------------|:---------------------|
| longName          | Apple Inc.           |
| industry          | Consumer Electronics |
| sector            | Technology           |
| marketCap         | 3805292789760        |
| sharesOutstanding | 14681140000          |

In [16]:
financial_statement

'### Quarterly Income Statement\n| 항목             |   2025-12-31 |   2025-09-30 |   2025-06-30 |   2025-03-31 |   2024-12-31 |\n|:-----------------|-------------:|-------------:|-------------:|-------------:|-------------:|\n| Total Revenue    |  1.43756e+11 |  1.02466e+11 |   9.4036e+10 |   9.5359e+10 |   1.243e+11  |\n| Gross Profit     |  6.9231e+10  |  4.8341e+10  |   4.3718e+10 |   4.4867e+10 |   5.8275e+10 |\n| Operating Income |  5.0852e+10  |  3.2427e+10  |   2.8202e+10 |   2.9589e+10 |   4.2832e+10 |\n| Net Income       |  4.2097e+10  |  2.7466e+10  |   2.3434e+10 |   2.478e+10  |   3.633e+10  |\n\n### Quarterly Balance Sheet\n| 항목                                    |   2025-12-31 |   2025-09-30 |   2025-06-30 |   2025-03-31 |   2024-12-31 |   2024-09-30 |\n|:----------------------------------------|-------------:|-------------:|-------------:|-------------:|-------------:|-------------:|\n| Total Assets                            |  3.79297e+11 |  3.59241e+11 |  3.31495e+11 |

### Quarterly Income Statement
| 항목             |   2025-12-31 |   2025-09-30 |   2025-06-30 |   2025-03-31 |   2024-12-31 |
|:-----------------|-------------:|-------------:|-------------:|-------------:|-------------:|
| Total Revenue    |  1.43756e+11 |  1.02466e+11 |   9.4036e+10 |   9.5359e+10 |   1.243e+11  |
| Gross Profit     |  6.9231e+10  |  4.8341e+10  |   4.3718e+10 |   4.4867e+10 |   5.8275e+10 |
| Operating Income |  5.0852e+10  |  3.2427e+10  |   2.8202e+10 |   2.9589e+10 |   4.2832e+10 |
| Net Income       |  4.2097e+10  |  2.7466e+10  |   2.3434e+10 |   2.478e+10  |   3.633e+10  |

### Quarterly Balance Sheet
| 항목                                    |   2025-12-31 |   2025-09-30 |   2025-06-30 |   2025-03-31 |   2024-12-31 |   2024-09-30 |
|:----------------------------------------|-------------:|-------------:|-------------:|-------------:|-------------:|-------------:|
| Total Assets                            |  3.79297e+11 |  3.59241e+11 |  3.31495e+11 |  3.31233e+11 |  3.44085e+11 |          nan |
| Total Liabilities Net Minority Interest |  2.91107e+11 |  2.85508e+11 |  2.65665e+11 |  2.64437e+11 |  2.77327e+11 |          nan |
| Stockholders Equity                     |  8.819e+10   |  7.3733e+10  |  6.583e+10   |  6.6796e+10  |  6.6758e+10  |          nan |

### Quarterly Cash Flow
| 항목                |   2025-12-31 |   2025-09-30 |   2025-06-30 |   2025-03-31 |   2024-12-31 |   2024-09-30 |
|:--------------------|-------------:|-------------:|-------------:|-------------:|-------------:|-------------:|
| Operating Cash Flow |   5.3925e+10 |   2.9728e+10 |   2.7867e+10 |   2.3952e+10 |   2.9935e+10 |          nan |
| Investing Cash Flow |  -4.886e+09  |  -2.587e+09  |   5.073e+09  |   2.917e+09  |   9.792e+09  |          nan |
| Financing Cash Flow |  -3.9656e+10 |  -2.7476e+10 |  -2.4833e+10 |  -2.9006e+10 |  -3.9371e+10 |          nan |

In [17]:
investment_report(company, basic_info, financial_statement)

'# Apple Inc. 분석 보고서\n\n## 1) 사업 개요와 최근 분기 핵심 포인트 3가지\n- **사업 개요**: Apple Inc.는 소비자 전자제품을 주로 생산하는 기술 기업으로, 아이폰, 아이패드, 맥북, 애플워치 등 다양한 제품을 제공합니다. 또한, 소프트웨어 및 서비스 부문에서도 강력한 입지를 가지고 있습니다.\n- **핵심 포인트**:\n  1. **매출 증가**: 2025년 12월 분기 총 매출이 143.76억 달러로, 이전 분기 대비 큰 폭으로 증가하였습니다.\n  2. **순이익 증가**: 같은 분기 순이익이 42.10억 달러로, 이전 분기보다 52.9% 증가하여 긍정적인 실적을 보였습니다.\n  3. **운영 현금 흐름 개선**: 운영 현금 흐름이 53.93억 달러로 증가하여 기업의 현금 생성 능력이 강화되었습니다.\n\n## 2) 실적 추세 요약\n- **매출**: \n  - 2025-12-31: 143.76억 달러\n  - 2025-09-30: 102.47억 달러\n  - 2025-06-30: 94.04억 달러\n- **영업이익**: \n  - 2025-12-31: 50.85억 달러\n  - 2025-09-30: 32.43억 달러\n  - 2025-06-30: 28.20억 달러\n- **순이익**: \n  - 2025-12-31: 42.10억 달러\n  - 2025-09-30: 27.47억 달러\n  - 2025-06-30: 23.43억 달러\n\n## 3) 리스크 2가지, 모멘텀 2가지\n- **리스크**:\n  1. **글로벌 공급망 문제**: 반도체 및 부품 공급의 불안정성이 지속될 경우 생산과 판매에 영향을 미칠 수 있습니다.\n  2. **경쟁 심화**: 삼성, 구글 등 다른 기술 기업들과의 경쟁이 치열해지면서 시장 점유율에 부정적인 영향을 미칠 수 있습니다.\n  \n- **모멘텀**:\n  1. **서비스 부문 성장**: 애플의 서비스 매출이 증가하고 있어 지속적인 수익원으로 작용할 가능성이 높습니다.\n  2. **신

# Apple Inc. 분석 보고서
## 1) 사업 개요와 최근 분기 핵심 포인트 3가지
**사업 개요**: Apple Inc.는 소비자 전자제품을 주로 생산하는 기술 기업으로, 아이폰, 아이패드, 맥북, 애플워치 등 다양한 제품을 제공합니다. 또한, 소프트웨어 및 서비스 부문에서도 강력한 입지를 가지고 있습니다.

**핵심 포인트**:
- 매출 증가: 2025년 12월 분기 총 매출이 143.76억 달러로, 이전 분기 대비 큰 폭으로 증가하였습니다.
- 순이익 증가: 같은 분기 순이익이 42.10억 달러로, 이전 분기보다 52.9% 증가하여 긍정적인 실적을 보였습니다.
- 운영 현금 흐름 개선: 운영 현금 흐름이 53.93억 달러로 증가하여 기업의 현금 생성 능력이 강화되었습니다.

## 2) 실적 추세 요약
**매출**:
- 2025-12-31: 143.76억 달러
- 2025-09-30: 102.47억 달러
- 2025-06-30: 94.04억 달러    

**영업이익**:
- 2025-12-31: 50.85억 달러
- 2025-09-30: 32.43억 달러
- 2025-06-30: 28.20억 달러   

**순이익**:
- 2025-12-31: 42.10억 달러
- 2025-09-30: 27.47억 달러
- 2025-06-30: 23.43억 달러

## 3) 리스크 2가지, 모멘텀 2가지
**리스크**:
- 글로벌 공급망 문제: 반도체 및 부품 공급의 불안정성이 지속될 경우 생산과 판매에 영향을 미칠 수 있습니다.
- 경쟁 심화: 삼성, 구글 등 다른 기술 기업들과의 경쟁이 치열해지면서 시장 점유율에 부정적인 영향을 미칠 수 있습니다.  

**모멘텀**:
- 서비스 부문 성장: 애플의 서비스 매출이 증가하고 있어 지속적인 수익원으로 작용할 가능성이 높습니다.
- 신제품 출시: 새로운 제품 라인업과 혁신적인 기술이 소비자에게 긍정적인 반응을 얻을 수 있습니다.

## 4) 투자 아이디어와 관찰지표 체크리스트
**투자 아이디어**: Apple의 지속적인 매출 성장과 강력한 브랜드 충성도를 바탕으로 장기 투자에 적합한 종목으로 평가됩니다.  

**관찰지표 체크리스트**:
- 분기별 매출 및 순이익 성장률
- 서비스 부문 매출 비중
- 운영 현금 흐름 및 자산 대비 부채 비율
- 신제품 출시 일정 및 소비자 반응
- 글로벌 경제 동향 및 공급망 안정성    

## 5) 10줄 요약
Apple Inc.는 소비자 전자제품 분야의 선두주자로, 최근 분기에서 매출과 순이익이 크게 증가하였습니다. 2025년 12월 분기 총 매출은 143.76억 달러로, 이전 분기 대비 40.5% 증가하였고, 순이익은 42.10억 달러로 52.9% 증가했습니다. 운영 현금 흐름도 개선되어 기업의 재무 건전성이 강화되었습니다. 그러나 글로벌 공급망 문제와 경쟁 심화가 리스크로 작용할 수 있습니다. 반면, 서비스 부문 성장과 신제품 출시가 긍정적인 모멘텀으로 작용할 가능성이 높습니다. Apple은 장기 투자에 적합한 종목으로 평가되며, 매출 성장률과 서비스 매출 비중을 주의 깊게 관찰해야 합니다.